bahmnicore-api/src/test/java/org/bahmni/module/bahmnicore/forms2/mapper/FormDetailsMapperTest.java

```diff
@@ -0,0 +1,238 @@
+import org.mockito.internal.verification.VerificationModeFactory;
+import org.openmrs.Concept;
+import org.openmrs.ConceptName;
+import org.openmrs.Encounter;
+import org.openmrs.Obs;
+import org.openmrs.PersonName;
+import org.openmrs.User;
+import org.openmrs.Visit;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+import org.powermock.reflect.Whitebox;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Date;
+import java.util.HashSet;
+
+import static java.util.Collections.singletonList;
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.powermock.api.mockito.PowerMockito.mockStatic;
+import static org.powermock.api.mockito.PowerMockito.verifyStatic;
+
+@PrepareForTest({FormUtil.class, FormType.class})
+@RunWith(PowerMockRunner.class)
+public class FormDetailsMapperTest {
+
+    private String formFieldPath = "FormName.2/1-0";
+    private String encounterUuid = "encounter-Uuid";
+    private String visitUuid = "visitUuid";
+    private String providerName = "Super Man";
+    private String providerUuid = "provider-uuid";
+    private String formName = "formName";
+    private int formVersion = 2;
+    private Date encounterDateTime = new Date();
+    private Date visitStartDateTime = new Date();
+
+    private Obs obs = mock(Obs.class);
+    private Encounter encounter = mock(Encounter.class);
+    private Visit visit = mock(Visit.class);
+    private User anotherCreator = mock(User.class);
+    private PersonName personName = mock(PersonName.class);
+
+    @Before
+    public void setUp() {
+
+        mockStatic(FormUtil.class);
+
+        when(obs.getEncounter()).thenReturn(encounter);
+        when(obs.getCreator()).thenReturn(anotherCreator);
+        when(obs.getFormFieldPath()).thenReturn(formFieldPath);
+        when(encounter.getVisit()).thenReturn(visit);
+        when(FormUtil.getFormNameFromFieldPath(formFieldPath)).thenReturn(formName);
+        when(FormUtil.getFormVersionFromFieldPath(formFieldPath)).thenReturn(formVersion);
+
+        when(encounter.getUuid()).thenReturn(encounterUuid);
+        when(encounter.getEncounterDatetime()).thenReturn(encounterDateTime);
+
+        when(visit.getUuid()).thenReturn(visitUuid);
+        when(visit.getStartDatetime()).thenReturn(visitStartDateTime);
+
+        when(anotherCreator.getPersonName()).thenReturn(personName);
+        when(personName.getFullName()).thenReturn(providerName);
+        when(anotherCreator.getUuid()).thenReturn(providerUuid);
+    }
+
+    @Test
+    public void shouldReturnFormDetailsFromGivenObsAndFormTypeOfFormBuilder() {
+
+        FormType formType = mock(FormType.class);
+        Whitebox.setInternalState(FormType.class, "FORM_BUILDER_FORMS", formType);
```

Comment:
This must not be done. using reflection to change internal state, that too after mocking it first. What are you trying to solve, that can not be solved with proper test data and code?

Reply:
https://github.com/Bahmni/bahmni-core/pull/41#discussion_r252260345